### PR TITLE
Support https communication between nodes

### PR DESCRIPTION
**Pull request by arhimondr**

### CODE REVIEW

**Comment by cawallin** [comment]
9ac6310 is a workaround until https://github.com/prestodb/presto/pull/6410 gets in

File: presto-docs/src/main/sphinx/security/internal-communication.rst [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        http-server.http.enabled=false
+
+    .. warning::
+
+        You can enable HTTPS while leaving HTTP enabled. In most cases this is a
+        security hole. If you are certain you want to use this configuration, you
+        should consider using an firewall to limit access to the HTTP endpoint to
+        only those hosts that should be allowed to use it.
+
+2. Configure cluster to communicate using the fully qualified domain name (fqdn)
```
Comment (cawallin) [inline presto-docs/src/main/sphinx/security/internal-communication.rst]: nit: Configure the cluster

**Comment by kwondw** [comment]
Thank you Andrii for this work. This is really helpful to support more secure Presto cluster.
I have a quick question, since this patch set InternalCommunicationConfig to HttpClientConfig, which seem to be used in airlift http client.
I just wonder why you choose to set keystore instead of truststore? for 2 way SSL? or I missed something.

**Comment by arhimondr** [comment]
> I just wonder why you choose to set keystore instead of truststore?

Good question. I think we need support both. 

For our particular use case we need to support keystore (it is easier to just generate a certificate and distribute it across the cluster without any need to sign it by the authority).

**Comment by kwondw** [comment]
I see, thanks for the quick and detail answer. I agree, it would be nice to support both as well as keyManagerPassword for keystore.
I was working on supporting to have separate keystore.password and keymanager.password(key.password) for keystore, it turns out since InternalCommunicationConfig is used by http client, it wouldn't need accessing to private key. so if InternalCommunicationConfig supports truststore, 2 different passwords wouldn't need it at all for the keystore has separate password.
Although, I can just set truststore in the context of keystore config in InternalCommunicationConfig with one password(truststore password) which works, it would be nice to have separate settings.

File: pom.xml [hunk 2]
```diff
@@ -116,6 +116,14 @@
         <module>presto-benchto-benchmarks</module>
     </modules>
 
+    <repositories>
```
Comment (dain) [inline pom.xml]: I believe the required Airlift changes have been merged and released.  If so, can you remove this commit.  If not, please link to the Airlift PR so I can get it merged.

File: presto-main/src/main/java/com/facebook/presto/failureDetector/HeartbeatFailureDetector.java [hunk 3]
```diff
@@ -251,18 +256,17 @@ void updateMonitoredServices()
...
     }
 
-    private static URI getHttpUri(ServiceDescriptor service)
+    private URI getHttpUri(ServiceDescriptor descriptor)
     {
-        try {
-            String uri = service.getProperties().get("http");
-            if (uri != null) {
-                return new URI(uri);
+        String type = httpsRequired ? "https" : "http";
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/failureDetector/HeartbeatFailureDetector.java]: inline this variable

File: presto-main/src/main/java/com/facebook/presto/metadata/DiscoveryNodeManager.java [hunk 4]
```diff
@@ -323,16 +326,15 @@ public Node getCurrentNode()
...
     {
-        for (String type : asList("http", "https")) {
-            String url = descriptor.getProperties().get(type);
-            if (url != null) {
-                try {
-                    return new URI(url);
-                }
-                catch (URISyntaxException ignored) {
-                }
+        String type = httpsRequired ? "https" : "http";
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/metadata/DiscoveryNodeManager.java]: inline this variable

File: presto-docs/src/main/sphinx/security/internal-communication.rst [hunk 5]
```diff
@@ -112,3 +112,45 @@ To enable SSL/TLS for Presto internal communication, do the following:
...
+For queries that do not require transferring too much data between the Presto
+nodes (e.g. `SELECT count(*) FROM table`), the performance impact is negligible.
+
+However, for CPU intensive queries which require a considerable amount of data
+to be transferred between the nodes (for example, distributed joins, aggregations and
+window functions, which require repartitioning), the performance impact might be
+considerable. The slowdown may vary from 10% to even 100%+, depending on the network
+traffic and the CPU utilization.
+
+Advanced Performance Tuning
```
Comment (dain) [inline presto-docs/src/main/sphinx/security/internal-communication.rst]: As a followup, we should add a section about crypto suites.

File: presto-product-tests/bin/run_on_docker.sh [hunk 6]
```diff
@@ -189,6 +190,11 @@ if [[ "$ENVIRONMENT" == "multinode-tls" ]]; then
     CLI_ARGUMENTS="--server https://presto-master.docker.cluster:7778 --keystore-path /docker/volumes/conf/presto/etc/docker.cluster.jks --keystore-password 123456"
 fi
 
+TLS_CERTIFICATE=""
```
Comment (arhimondr) [inline presto-product-tests/bin/run_on_docker.sh]: Do you think it is possible to merge https://github.com/prestodb/presto/pull/7725 before this PR? So we don't need to add this workaround.

Comment (arhimondr) [inline presto-product-tests/bin/run_on_docker.sh]: @electrum ?

File: presto-docs/src/main/sphinx/security/internal-communication.rst (hunk 5 above)
Comment (arhimondr) [inline presto-docs/src/main/sphinx/security/internal-communication.rst]: What do you mean? Do you think we should add a property where the transfer encryption algorithm might be changed?

Comment (dain) [inline presto-docs/src/main/sphinx/security/internal-communication.rst]: The most recent crypto suites are very expensive in Java (we saw 90% cpu usage for them in production), so we should note this is something users should watch out for, and maybe suggest specific suites.

**Comment by dain** [comment]
@arhimondr, I'll merge this now and take a look at https://github.com/prestodb/presto/pull/7725 soonish.

**Comment by dain** [comment]
Merged, thanks!

**Pull request closed without merging**